flink-scala/src/main/scala/org/apache/flink/api/scala/typeutils/ScalaEnumSerializerSnapshot.scala

```diff
@@ -0,0 +1,136 @@
+
+      out.writeInt(previousEnumConstants.length)
+      for ((name, idx) <- previousEnumConstants) {
+        out.writeUTF(name)
+        out.writeInt(idx)
+      }
+    } finally if (outViewWrapper != null) outViewWrapper.close()
+  }
+
+  override def readSnapshot(
+    readVersion: Int, in: DataInputView, userCodeClassLoader: ClassLoader): Unit = {
+    val inViewWrapper = new DataInputViewStream(in)
+    try {
+      if (readVersion == 1) {
+        enumClass = InstantiationUtil.deserializeObject(
+          inViewWrapper, userCodeClassLoader)
+
+        // read null from input stream
+        InstantiationUtil.deserializeObject(inViewWrapper, userCodeClassLoader)
+        previousEnumConstants = List()
+      } else if (readVersion >= 2) {
+        enumClass = Class.forName(
+          in.readUTF(), true, userCodeClassLoader).asInstanceOf[Class[E]]
+
+        val length = in.readInt()
+        val listBuffer = ListBuffer[(String, Int)]()
+
+        for (_ <- 0 until length) {
+          val name = in.readUTF()
+          val idx = in.readInt()
+          listBuffer += ((name, idx))
+        }
+
+        previousEnumConstants = listBuffer.toList
+      } else {
+        throw new IOException(
+          s"Cannot deserialize ${getClass.getSimpleName} with version $readVersion.")
+      }
+    } catch {
+      case e: ClassNotFoundException =>
+        throw new IOException("The requested enum class cannot be found in classpath.", e)
+    }
+    finally if (inViewWrapper != null) inViewWrapper.close()
+  }
+
+  override def restoreSerializer(): TypeSerializer[E#Value] = {
+    enumClass.newInstance().asInstanceOf[TypeSerializer[E#Value]]
+  }
+
+  override def resolveSchemaCompatibility(
+    newSerializer: TypeSerializer[E#Value]): TypeSerializerSchemaCompatibility[E#Value] = {
+    newSerializer match {
+      case newEnumSerializer: EnumValueSerializer[_] => {
+        if (enumClass.equals(newEnumSerializer.enum.getClass)) {
+          for ((previousEnumConstant, idx) <- previousEnumConstants) {
+            val enumValue = try {
+              newEnumSerializer.enum(idx)
+            } catch {
+              case _: NoSuchElementException =>
+                // couldn't find an enum value for the given index
+                return TypeSerializerSchemaCompatibility.incompatible()
+            }
+
+            if (!previousEnumConstant.equals(enumValue.toString)) {
+              // compatible only if new enum constants are only appended,
+              // and original constants must be in the exact same order
+              return TypeSerializerSchemaCompatibility.compatibleAfterMigration()
+            }
+          }
+
+          TypeSerializerSchemaCompatibility.compatibleAsIs()
+        } else {
+          TypeSerializerSchemaCompatibility.incompatible()
+        }
+      }
+      case _ => TypeSerializerSchemaCompatibility.incompatible()
+    }
+  }
+}
+
+object ScalaEnumSerializerSnapshot {
+  val VERSION = 3
```

Comment:
To be in line with the other newly added `TypeSerializerSnapshots`, this should maybe be version `2`. But I see that version `3` could also be valid since the old `TypeSerializerConfigSnapshot` for enums was already at version `2`.

@igalshilman Sorry for bothering again, but what do you think?